### DISCUSSION TITLE
stubPath for matplotlib help

### DISCUSSION THREAD

**Discussion by jdtsmith** [microsoft/pyright#4240] [Q&A] [answered]
I've downloaded [python-type-stubs](https://github.com/microsoft/python-type-stubs), and linked its `matplotlib` directory into a common directory, for all my projects to use.  Then I send configuration to pyright, ala:

```
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
	  (:settings
	   (:python
	    (:stubPath "../../code/python/stubs/common"))))
```

pyright knows about the module, as it completes the import (from my default user environment, if it matters).  But no matter what I try, I can't get completion for pyplot:

```python
from matplotlib import pyplot as plt
plt.fi[Tab]  # fi: Unknown :(
```

I've tried every variation for `stubPath`: absolute, relative, with and without the final `/`.   The only way I can get it to work is to link it into `./typings` and remove my `stubPath` config.  Is it not allowed to have `stubPath` point outside the workspace or something?  How do I install stubs "globally", for packages that are in the system or user environment?

**Comment by erictraut** [comment, marked as answer]
It looks like you're specifying the wrong setting. It's "python.analysis.stubPath", not "python.stubPath".

**Reply by jdtsmith** [reply]
Thanks, that's important.  Even correcting this, I still can't seem to get pyright to take the setting. What's strange is that python.pythonPath works fine using this method, but python.analysis.stubPath is apparently silently ignored after the server does a workspace/configuration request and gets the results.  With pythonPath I get a window/logMessage of "Setting pythonPath for service...".  Should there be a similar log item for analysis.stubPath?  Current config getting sent looks like:

```
(:settings
	   (:python
	    (:analysis
	     (:stubPath "./footype")
	     :pythonPath "/Users/.../code/python/pahfit/.venv/bin/python3"))))
```

(testing with a local "./footype" directory).

**Reply by erictraut** [reply]
No, there is not a similar log item for analysis.stubPath.

Why are you manually crafting RPC messages to send to the language server? That sounds very error prone. Doesn't your client support a settings mechanism? I recommend using that rather than manually formatting your own RPC messages.

You can also specify the stubPath in a pyrightconfig.json or pyproject.toml configuration file.

What you're doing here is outside the intended usage of pyright, so you may need to do your own debugging to diagnose the issue. See [this documentation](https://github.com/microsoft/pyright/blob/main/docs/build-debug.md) for building and debugging pyright.

**Reply by jdtsmith** [reply]
This is emacs with eglot (soon to be the in-built LSP client).  Its settings mechanism is rather primitive: the user just provides a lisp structure that, when converted to JSON, the server will understand.  Here I'm simply reporting the traffic that eglot logs, e.g. with its `workspace/didChangeConfiguration` method (eglot quirkily displays the JSON it sends as equivalent lisp lists).

I did figure it out.  My apologies, this is entirely an eglot oddity.  When the server requests the `python.analysis` config section, eglot doesn't process that hierarchically, but simply looks for a key matching a literal `:python.analysis`.   Once I corrected for this, it works great.  

It do understand it's a challenge for a general LSP client like eglot to work with settings, since servers employ and interpret them in a variety of ways.  Not sure how many levels of "settings" hierarchy the LSP standard specifies, but this could likely be considered an eglot bug; will report. 

Appreciate your work.

**Comment by lld2001** [comment]
@jdtsmith  Could you please share your `eglot-workspace-configuration` config? The same issue I found.

**Reply by jdtsmith** [reply]
Sure: 

```elisp
(defun my/eglot-workspace-config (server)
    (let ((config (list :python.analysis ; N.B. eglot does not consider :settings nested!
			(list :stubPath (expand-file-name "~/code/python/stubs/common")))))
      (when-let ((venv (simple-venv-find))
		 ((not (eq venv 'none)))
		 (vp (file-name-directory venv))
		 (vn (file-name-nondirectory venv)))
	(nconc config (list :python
			    `(:venvPath ,vp :venv ,vn
			     :pythonPath ,(simple-venv-interpreter venv)))))
      config))

  (setq-default eglot-workspace-configuration #'my/eglot-workspace-config)
```

This makes use of an unreleased mode for finding enclosing venv, but you can get the idea.

**Reply by lld2001** [reply]
I see. Thank you so much!

It seems this mode work well with eglot.
